src/Native/System.Private.CoreLib.Native/pal_time.cpp

```diff
@@ -33,3 +34,161 @@ mach_timebase_info_data_t *InitializeTimebaseInfo()
     return g_isMachTimebaseInfoInitialized ? &g_machTimebaseInfo : nullptr;
 }
 #endif
+
+extern "C" uint64_t CoreLibNative_GetHighPrecisionCounts()
+{
+    uint64_t counts = 0;
+
+#if HAVE_MACH_ABSOLUTE_TIME
+    {
+        counts = mach_absolute_time();
+    }
+#elif HAVE_CLOCK_MONOTONIC_COARSE || HAVE_CLOCK_MONOTONIC
+    {
+        clockid_t clockType =
+#if HAVE_CLOCK_MONOTONIC_COARSE
+            CLOCK_MONOTONIC_COARSE; // good enough resolution, fastest speed
+#else
+            CLOCK_MONOTONIC;
+#endif
+        struct timespec ts;
+        if (clock_gettime(clockType, &ts) != 0)
+        {
+            assert(false);
+            return counts;
+        }
+        counts = ts.tv_sec * NanosecondsPerSecond + ts.tv_nsec;
+    }
+#else
+    {
+        struct timeval tv;
+        if (gettimeofday(&tv, nullptr) == -1)
+        {
+            assert(false);
+            return counts;
+        }
+        counts = tv.tv_sec * MicrosecondsPerSecond + tv.tv_usec;
+    }
+#endif
+    return counts;
+}
+
+extern "C" uint64_t CoreLibNative_GetHighPrecisionCounterFrequency()
+{
+    static uint64_t frequency = 0;
```

Comment:
It would be better to make this a static global in this file, no need for lazy initialization of this